packages/components/src/components/SquiggleViewer/utils.ts

```diff
@@ -104,3 +109,103 @@ function isHidden(value: SqValue): boolean {
   return value.entries().filter(([_, v]) => !isHidden(v));
 }
+
+function valueToLine(value: SqValue): number | undefined {
+  return value.context?.findLocation()?.start.line;
+}
+
+//This doesn't get imports, because their contexts are wrong.
+export function findValuePathByLine(
+  line: number,
+  sqResult?: SqOutputResult
+): { type: "Imports" | "Variables" | "Result"; path: SqValuePath } | undefined {
+  if (!sqResult?.ok) {
+    return undefined;
+  }
+  const { bindings, result } = sqResult.value;
+
+  function allChildren(value: SqValue): SqValue[] {
+    return getChildrenValues(value).flatMap((v) => [v, ...allChildren(v)]);
+  }
+
+  // Search in bindings
+  for (const value of allChildren(bindings.asValue())) {
+    if (line === valueToLine(value) && value.context?.path) {
+      return { type: "Variables", path: value.context.path };
+    }
+  }
+
+  // Check in result
+  const resultLine = valueToLine(result);
+  if (line === resultLine && result.context?.path) {
+    return { type: "Result", path: result.context.path };
+  }
+}
+
+function lastUniqBy<A, B>(arr: A[], fn: (a: A) => B): A[] {
+  return uniqBy(arr.reverse(), fn).reverse();
+}
+
+function visibleAstChildren(node: ASTNode): ASTNode[] {
+  switch (node.type) {
+    case "Dict":
+      // Assuming 'elements' is an array of { key: ASTNode, value: ASTNode | string }
+      // and we only want to include ASTNode values
+      return node.elements.flatMap((e) =>
+        typeof e.value === "string" ? [] : [e.value]
+      );
+    case "Array":
+      return node.elements;
+    case "Block": {
+      const lastNode = node.statements.at(-1);
+      return lastNode ? [lastNode] : [];
+    }
+    case "LetStatement":
+      return [node.value];
+    case "DecoratedStatement":
+      return node.decorator.type === "Decorator" &&
+        node.decorator.name.value !== "hide"
+        ? [node.statement]
+        : [];
+    case "Program": {
+      return lastUniqBy(node.statements, (s) => {
+        switch (s.type) {
+          case "LetStatement":
+            return s.variable.value;
+          case "Identifier":
+            return s.value;
+          default:
+            return s;
+        }
+      });
+    }
+    default:
+      return [];
+  }
+}
+
+function visibleAstLeaves(node: ASTNode): ASTNode[] {
+  return [node, ...visibleAstChildren(node).flatMap(visibleAstLeaves)];
+}
+
+//This doesn't get imports, because their contexts are wrong.
+export function getActiveLineNumbers(sqResult?: SqOutputResult): number[] {
+  if (!sqResult?.ok) {
+    return [];
+  }
+  const ast = sqResult.value.bindings.context?.ast;
+  if (!ast) {
+    return [];
+  }
+  const values = uniq(
+    visibleAstLeaves(ast)
+      .filter((p) =>
+        includes(
+          ["LetStatement", "DefunStatement", "Block", "Dict", "Identifier"],
+          p.type
+        )
+      )
+      .map((s) => s.location.start.line - 1)
+  );
+  return values;
```

Comment:
Looking back, I don't think this refactor to use the AST was worth it yet - mainly because we're not even getting to use the AST pre-render anyway. This code will be a bit more buggy than going straight to values, as it won't always align perfectly.